### PR TITLE
Put SPMediaKeyTap back

### PR DESCRIPTION
**Pull request by lassana**
Relates to #607. With these changes iTunes/Apple Music not longer launches when the Play/Pause button key is pressed. This trick works only after user manually enables accessibility access permission for strawberry:

Screen Shot 2020-12-27 at 20 50 05

It would be nice to show the Accessibility Access request window but I'm not really sure how to do that. It's worth mentioning that this dialog is already shown every time when the Preferences window is opened (until user enables it or selects "Deny"):

Screen Shot 2020-12-28 at 00 54 22

Once the accessibly permission is given it's required to restart the app. Ideally Strawberry would have a macOS-specific preference for this feature and start/stop media key monitoring when user enables/disables it.

SPMediaKeyTap is from the latest master (338bebf).

This PR, however, does not fix an issue when iTunes/Apple Music starts when user presses the _headset_ Play/Pause button.

### CODE REVIEW

**Comment by jonaski** [comment]
I think the SPMediaKeyTap in VLC is more up-to-date. I see it has fixed the semicolon issue, and does not use some of the deprecated methods such as GetCurrentProcess, SameProcess and ProcessInformationCopyDictionary
https://github.com/videolan/vlc/tree/master/modules/gui/macosx/imported/SPMediaKeyTap
Maybe we could switch to their version, if that works. Does not look like SPInvocationGrabbing and SPMediaKeyTapDelegate.m is needed either.

**Comment by lassana** [comment]
Alright, I have set up Qt Creator dev environment and will try to adapt VLC code.

**Comment by jonaski** [comment]
If it doesn't work, and if you don't know how to replace use of the deprecated functions you can use  `#pragma GCC diagnostic ignored "-Wdeprecated-declarations"` like I did here: https://github.com/strawberrymusicplayer/strawberry/commit/47fd30c23fce80c662acf1a55d05ca73bdf7fdde
so the CI passes. But preferable we should try replace use of it now.

**Comment by lassana** [comment]
I've copied SPMediaKeyTap from VLC, it seems to work well but Strawberry crashes every time when it comes from background. Maybe someone with better knowledge of macOS will be able to figure this out.

<details>

```
Crashed Thread:        0  Dispatch queue: com.apple.main-thread

Exception Type:        EXC_BAD_ACCESS (SIGSEGV)
Exception Codes:       KERN_INVALID_ADDRESS at 0x0000005400000018
Exception Note:        EXC_CORPSE_NOTIFY

Termination Signal:    Segmentation fault: 11
Termination Reason:    Namespace SIGNAL, Code 0xb
Terminating Process:   exc handler [49353]

VM Regions Near 0x5400000018:
    VM_ALLOCATE (reserved) 000000013b030000-000000013b038000 [   32K] rw-/rwx SM=NUL  reserved VM address space (unallocated)
--> 
    MALLOC_NANO            0000600000000000-0000600008000000 [128.0M] rw-/rwx SM=PRV  

Application Specific Information:
objc_msgSend() selector name: containsObject:


Thread 0 Crashed:: Dispatch queue: com.apple.main-thread
0   libobjc.A.dylib               	0x00007fff676d581d objc_msgSend + 29
1   org.strawberrymusicplayer.strawberry	0x000000010d75ced2 0x10cf80000 + 8244946
2   com.apple.CoreFoundation      	0x00007fff2e7e980f __CFNOTIFICATIONCENTER_IS_CALLING_OUT_TO_AN_OBSERVER__ + 12
3   com.apple.CoreFoundation      	0x00007fff2e7e97a3 ___CFXRegistrationPost1_block_invoke + 63
4   com.apple.CoreFoundation      	0x00007fff2e7e9718 _CFXRegistrationPost1 + 372
5   com.apple.CoreFoundation      	0x00007fff2e7e9384 ___CFXNotificationPost_block_invoke + 80
6   com.apple.CoreFoundation      	0x00007fff2e7b94fd -[_CFXNotificationRegistrar find:object:observer:enumerator:] + 1554
7   com.apple.CoreFoundation      	0x00007fff2e7b89a9 _CFXNotificationPost + 1351
8   com.apple.Foundation          	0x00007fff30e36746 -[NSNotificationCenter postNotificationName:object:userInfo:] + 59
9   com.apple.AppKit              	0x00007fff2bc9d2c6 applicationStatusSubsystemCallback + 736
10  com.apple.LaunchServices      	0x00007fff2ff4986e ___LSScheduleNotificationFunction_block_invoke_2 + 47
11  com.apple.CoreFoundation      	0x00007fff2e7f39ec __CFRUNLOOP_IS_CALLING_OUT_TO_A_BLOCK__ + 12
12  com.apple.CoreFoundation      	0x00007fff2e7f3934 __CFRunLoopDoBlocks + 379
13  com.apple.CoreFoundation      	0x00007fff2e7f2e2d __CFRunLoopRun + 2450
14  com.apple.CoreFoundation      	0x00007fff2e7f1e3e CFRunLoopRunSpecific + 462
15  com.apple.HIToolbox           	0x00007fff2d41eabd RunCurrentEventLoopInMode + 292
16  com.apple.HIToolbox           	0x00007fff2d41e7d5 ReceiveNextEventCommon + 584
17  com.apple.HIToolbox           	0x00007fff2d41e579 _BlockUntilNextEventMatchingListInModeWithFilter + 64
18  com.apple.AppKit              	0x00007fff2ba64039 _DPSNextEvent + 883
19  com.apple.AppKit              	0x00007fff2ba62880 -[NSApplication(NSEvent) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 1352
20  com.apple.AppKit              	0x00007fff2ba5458e -[NSApplication run] + 658
21  libqcocoa.dylib               	0x00000001134fbd74 0x1134c0000 + 245108
22  org.qt-project.QtCore         	0x000000010f4a96b7 QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) + 471
23  org.qt-project.QtCore         	0x000000010f4ad852 QCoreApplication::exec() + 130
24  org.strawberrymusicplayer.strawberry	0x000000010cf9019d main + 7709
25  libdyld.dylib                 	0x00007fff6888acc9 start + 1
```

<details>

**Comment by jonaski** [comment]
It crashes in frontmostAppChanged on the following:
```
if (![[SPMediaKeyTap mediaKeyUserBundleIdentifiers] containsObject:app.bundleIdentifier])
return;
```
The old SPMediaKeyTap does not have frontmostAppChanged.

File: 3rdparty/SPMediaKeyTap/SPMediaKeyTap.m [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+        _mediaKeyAppList = [NSMutableArray new];
+    }
+    return self;
+}
+
+- (void)dealloc
+{
+    [self stopWatchingMediaKeys];
+    [self stopWatchingAppSwitching];
+}
```
Comment (jonaski) [inline 3rdparty/SPMediaKeyTap/SPMediaKeyTap.m:75]: Add `[super dealloc];`

File: 3rdparty/SPMediaKeyTap/SPMediaKeyTap.m [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+
+
+#pragma mark -
+#pragma mark Event tap callbacks
+
+// Note: method called on background thread
+
+static CGEventRef tapEventCallback2(CGEventTapProxy proxy, CGEventType type, CGEventRef event, void *refcon)
+{
```
Comment (jonaski) [inline 3rdparty/SPMediaKeyTap/SPMediaKeyTap.m:253]: `#pragma unused(proxy)`

**Comment by jonaski** [comment]
Removing this check in frontmostAppChanged fixes the crash, so it has something to do with the array.

```
if (![[SPMediaKeyTap mediaKeyUserBundleIdentifiers] containsObject:app.bundleIdentifier])
return;

```

**Comment by jonaski** [comment]
Removing the dispatch_once stuff and changing mediaKeyUserBundleIdentifiers to just return a NSArray like the old SPMediaKeyTap did seem to fix it. Example:
```
+ (NSArray*)mediaKeyUserBundleIdentifiers {
  return [NSArray arrayWithObjects:[[NSBundle mainBundle] bundleIdentifier], nil];
}
```
I'm not a objective C/C++ programmer so I'm not familiar with the dispatch_once stuff.

**Review by jonaski — CHANGES_REQUESTED** [review]
Change mediaKeyUserBundleIdentifiers to return a NSArray directly so it doesn't crash.
Like:
```
+ (NSArray*)mediaKeyUserBundleIdentifiers {
  return [NSArray arrayWithObjects:[[NSBundle mainBundle] bundleIdentifier], nil];
}
```

**Comment by jonaski** [comment]
@lassana 
Do you want to finish this?
If you make my suggested changes we can merge this

**Comment by lassana** [comment]
@jonaski sorry I didn't have much time recently but I'll definitely apply suggested changes soon™️. I think you could close this PR, I'll open another one when it will be ready.

File: src/core/mac_startup.mm [hunk 3]
```diff
@@ -171,6 +186,25 @@ - (void)application:(NSApplication*)app openFiles:(NSArray*)filenames {
 
 }
 
+- (void) mediaKeyTap: (SPMediaKeyTap*)keyTap receivedMediaKeyEvent:(NSEvent*)event {
```
Comment (jonaski) [inline src/core/mac_startup.mm:189]: Add `#pragma unused(keyTap)` to silence this warning on Mojave:
```
/Users/travis/build/strawberrymusicplayer/strawberry/src/core/mac_startup.mm:189:39: error: unused parameter 'keyTap' [-Werror,-Wunused-parameter]
- (void) mediaKeyTap: (SPMediaKeyTap*)keyTap receivedMediaKeyEvent:(NSEvent*)event {
```

**Review by jonaski — CHANGES_REQUESTED** [review]

**Review by jonaski — APPROVED** [review]

**Pull request merged**